### PR TITLE
etcdctl/ctlv3: Cleanup keys on signal interrupt.

### PR DESCRIPTION
**Pull request by lilic**
If ctc+c is pressed during the run of etcdctl check perf, this deletes all the previously created keys that perf check created.

I also cleaned the context a bit for this to be possible, so that we can cancel all the contexts before deleting, otherwise during deletion the new keys would keep being created.

### CODE REVIEW

File: etcdctl/ctlv3/command/error.go [hunk 1]
```diff
@@ -38,5 +40,8 @@ func ExitWithError(code int, err error) {
 	if cerr, ok := err.(*client.ClusterError); ok {
 		fmt.Fprintln(os.Stderr, cerr.Detail())
 	}
+	if errors.Is(err, context.Canceled) {
+		os.Exit(0)
```
Comment (ptabor) [inline etcdctl/ctlv3/command/error.go:44]: It's not very intuitive that some of the error codes are ignoring the provided 'code'. 

Maybe it should be sorted out in the upper layer ?

File: etcdctl/ctlv3/command/check.go [hunk 2]
```diff
@@ -157,13 +158,34 @@ func newCheckPerfCommand(cmd *cobra.Command, args []string) {
...
 	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(cfg.duration)*time.Second)
-	resp, err := clients[0].Get(ctx, checkPerfPrefix, v3.WithPrefix(), v3.WithLimit(1))
-	cancel()
+	defer cancel()
+
+	signalChan := make(chan os.Signal, 1)
+	defer signal.Stop(signalChan)
+	signal.Notify(signalChan, os.Interrupt)
+
+	go func() {
```
Comment (ptabor) [inline etcdctl/ctlv3/command/check.go:167]: Maybe move to a separate context: 

Ideally the integration could be based on invocation in following style: 

```
ctx := interruptableContext(ctx, func() { attemptCleanup(clients[0]) } )
```

Comment (lilic) [reply etcdctl/ctlv3/command/check.go]: Thanks for the great suggestion, I think I implemented what you had in mind, please have a look, thank you!

File: etcdctl/ctlv3/command/error.go (hunk 1 above)
Comment (lilic) [reply etcdctl/ctlv3/command/error.go:44]: If by upper layer you mean `newCheckPerfCommand` or `NewCheckPerfCommand`?

File: etcdctl/ctlv3/command/check.go [hunk 3]
```diff
@@ -265,6 +270,30 @@ func newCheckPerfCommand(cmd *cobra.Command, args []string) {
 	}
 }
 
+func attemptCleanup(client *v3.Client) {
+	dctx, dcancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer dcancel()
+	_, err := client.Delete(dctx, checkPerfPrefix, v3.WithPrefix())
+	if err != nil {
+		ExitWithError(ExitError, err)
```
Comment (lilic) [inline etcdctl/ctlv3/command/check.go:278]: I think we should probably just log here that delete failed and not exit.

Comment (ptabor) [reply etcdctl/ctlv3/command/check.go:278]: SGTM

File: etcdctl/ctlv3/command/check.go [hunk 4]
```diff
@@ -212,9 +217,9 @@ func newCheckPerfCommand(cmd *cobra.Command, args []string) {
 
 	s := <-sc
 
-	ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
-	dresp, err := clients[0].Delete(ctx, checkPerfPrefix, v3.WithPrefix())
-	cancel()
+	dctx, dcancel := context.WithTimeout(context.Background(), 30*time.Second)
```
Comment (ptabor) [inline etcdctl/ctlv3/command/check.go:220]: Maybe calling here  attemptCleanup()

File: etcdctl/ctlv3/command/check.go (hunk 2 above)
Comment (ptabor) [reply etcdctl/ctlv3/command/check.go]: Thank you for fixing.

File: etcdctl/ctlv3/command/check.go [hunk 5]
```diff
@@ -265,6 +270,30 @@ func newCheckPerfCommand(cmd *cobra.Command, args []string) {
...
+	signal.Notify(signalChan, os.Interrupt)
+	go func() {
+		defer signal.Stop(signalChan)
+		select {
+		case <-signalChan:
+			cancel()
+			attemptCleanup()
+		}
+	}()
+	return ctx
```
Comment (ptabor) [inline etcdctl/ctlv3/command/check.go:294]: Please return cancel here, as:

https://golang.org/pkg/context/#WithCancel

states: 
> Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete.

so we need to call cancel even if there was no 'signal.

File: etcdctl/ctlv3/command/error.go (hunk 1 above)
Comment (ptabor) [reply etcdctl/ctlv3/command/error.go:44]: I think that `attemptCleanup` would be good enough to expose 'cancellation' as os.Exit(0). 
Logging error there would also work.

File: etcdctl/ctlv3/command/check.go (hunk 4 above)
Comment (lilic) [reply etcdctl/ctlv3/command/check.go]: Yes nice, done!

**Comment by lilic** [comment]
Thanks @ptabor for the great comments, this should be ready for another round of reviews, thank you!

**Review by ptabor — APPROVED** [review]
Thank you for fixing it.

**Pull request merged**